components/ca-form.js

```diff
@@ -0,0 +1,1520 @@
+                    }, err => {
+                        trace('Error creating new record', err);
+                    });
+            } else {
+                this.onsave.call(this, formData); // @todo ... this method does not appear in weblab-common-ui...
+            }
+
+            return true;
+
+        }
+
+        /**
+         * @description Event handler fired when a record is created (return true to continue with next step)
+         * @param {object} data - json data that was sent to the server
+         * @param {object} next - next hateoas link returned from creation (if present)
+         * @returns {boolean} true always
+         */
+        oncreate() {
+            return true;
+        }
+
+        /**
+         * @description Event handler fired when a record is update (return true to continue with next step)
+         * @param {object} data - json data that was sent to the server
+         * @param {object} next - next hateoas link returned from update (if present)
+         * @returns {boolean} true always
+         */
+        onupdate() {
+            return true;
+        }
+
+        /**
+         * @description is an empty function
+         * @param {string} label describing item
+         * @returns {void} nothing nada
+         */ // eslint-disable-next-line no-unused-vars
+        onbuttonclick(label) {
+
+        }
+
+        /**
+         * @description is an empty function
+         * @returns {void} nothing nada
+         */
+        onrendercomplete() {
+
+        }
+
+        /**
+         * @description is an empty function
+         * @returns {void} nothing nada
+         */
+        onpopulatecomplete() {
+
+        }
+
+        /**
+         * @description Event handler fired when a schema loads
+         * @returns {void} nothing nada
+         */
+        onschemaloaded() {
+
+        }
+
+        /**
+         * @description Go through the data, for each key get the element and set it's value based on element type
+         * @param {object} data - data to bind to the form (defaults to internal data value)
+         * @returns {void} nothing
+         */
+        populate(data) {
+
+            // eslint-disable-next-line no-param-reassign
+            data = data || this.data;
+
+            // eslint-disable-next-line no-restricted-syntax
+            for (const key in data) {
+                if (key !== 'links' && key.indexOf('$') === -1 && Object.prototype.hasOwnProperty.call(data, key)) {
+
+                    const el = this.querySelector(`[name="${key}"]`);
+                    const value = (typeof data[key] !== 'undefined') ? data[key] : '';
+
+                    if (!el) {
+                        // eslint-disable-next-line no-continue
+                        continue;
+                    }
+
+                    setElementValue(el, value);
+                }
+            }
+
+            // fire onpopulatecomplete event
+            if (this.onpopulatecomplete) {
```

Comment:
pointless if statement, especially considering the pointless blank function exists